modules/library/data/src/main/java/org/geotools/data/store/ContentFeatureStore.java

```diff
@@ -283,37 +278,41 @@ public ContentFeatureStore(ContentEntry entry,Query query) {
         return writer;
     }
 
-    FeatureId addFeature(SimpleFeature feature, FeatureWriter<SimpleFeatureType, SimpleFeature> writer) throws IOException {
+    FeatureId addFeature(final SimpleFeature feature, FeatureWriter<SimpleFeatureType,
+            SimpleFeature> writer) throws IOException {
         // grab next feature and populate it
         // JD: worth a note on how we do this... we take a "pull" approach 
         // because the raw schema we are inserting into may not match the 
         // schema of the features we are inserting
-        SimpleFeature toWrite = writer.next();
+        final SimpleFeature toWrite = writer.next();
         for ( int i = 0; i < toWrite.getType().getAttributeCount(); i++ ) {
             String name = toWrite.getType().getDescriptor(i).getLocalName();
             toWrite.setAttribute( name, feature.getAttribute(name));
         }
         
         // copy over the user data
-        if(feature.getUserData().size() > 0) {
-            toWrite.getUserData().putAll(feature.getUserData());
-        }
-        
+        toWrite.getUserData().putAll(feature.getUserData());
+
         // pass through the fid if the user asked so
         boolean useExisting = Boolean.TRUE.equals(feature.getUserData().get(Hints.USE_PROVIDED_FID));
         if(getQueryCapabilities().isUseProvidedFIDSupported() && useExisting) {
             ((FeatureIdImpl) toWrite.getIdentifier()).setID(feature.getID());
         }
-        
+
+        //PV: I'm not very happy about this hack. But that's the only way I've found, with the
+        //    current infra, to have the feature's ID to be updated in case of delayed write
```

Comment:
Infrastructure.

I spent a long time trying to understand why the features are copied in the JDBCInsertFeatureWriter before inserting them. There is, somewhere a big comment explaining that the feature's structure may not match the DB structure, but why would that matter? If they don't match too much it won't work anyway.

That comment was more a note for the reviewer to pay attention to the hack and make sure it doesn't go un-noticed. Mission accomplished  ;-)

Now, do you see a better way? If not and you are OK with it, I'll remove the negativity and the personal touch from this comment.

By the way, thanks for the review.